neuro_san/internals/journals/progress_journal.py

```diff
@@ -0,0 +1,53 @@
+
+# Copyright (C) 2023-2025 Cognizant Digital Business, Evolutionary AI.
+# All Rights Reserved.
+# Issued under the Academic Public License.
+#
+# You can be released from the terms, and requirements of the Academic Public
+# License by purchasing a commercial license.
+# Purchase of a commercial license is mandatory for any use of the
+# neuro-san SDK Software in commercial settings.
+#
+# END COPYRIGHT
+from typing import Any
+from typing import Dict
+
+from neuro_san.interfaces.agent_network_progress_reporter import AgentNetworkProgressReporter
+from neuro_san.internals.journals.journal import Journal
+from neuro_san.internals.messages.agent_progress_message import AgentProgressMessage
+
+
+class ProgressJournal(AgentNetworkProgressReporter):
+    """
+    An implementation of the AgentNetworkProgressReporter interface for a CodedTool to be able
+    to journal AgentProgressMessages.
+    """
+
+    def __init__(self, wrapped_journal: Journal):
+        """
+        Constructor
+
+        :param wrapped_journal: The Journal that this implementation wraps
+        """
+        self.wrapped_journal: Journal = wrapped_journal
+
+    async def async_report_progress(self, structure: Dict[str, Any], content: str = ""):
+        """
+        Reports the structure and optional message to the chat message stream returned to the client
+        To be used from within CodedTool.async_invoke().
+
+        :param structure: The Dictionary instance to write as progress.
+                        All keys and values must be JSON-serializable.
+        :param content: An optional message to send to the client
+        """
```

Comment:
Worth noting that we create the AgentProgressMessage in here as an implementation detail.
This is also helpful to the client code since they'd just be creating the structure over and over again themselves if this wasn't here.